cpg_infra/abstraction/metamist.py

```diff
@@ -0,0 +1,99 @@
+"""
+Pulumi provider for interacting with metamist
+"""
+import functools
+import pulumi
+from metamist.apis import ProjectApi
+
+
+@functools.cache
+def get_projects() -> dict[str, dict]:
+    """
+    Get all projects from metamist, useful to avoid repeated calls to the API
+    :return:
+    """
+    api = ProjectApi()
+    all_projects = api.get_all_projects()
+    return {p['name']: p for p in all_projects}
+
+
+def get_project_by_name(name: str) -> dict | None:
+    """
+    Get a project by name from metamist, uses the cached get_projects() function
+    """
+    projects = get_projects()
+    return projects.get(name)
+
+
+class MetamistProjectProvider(pulumi.dynamic.ResourceProvider):
+    """Pulumi provider for creating a metamist project"""
+
+    def create(self, props) -> pulumi.dynamic.CreateResult:
+        name = props['project_name']
+        create_test_project = props['create_test_project']
+
+        existing_project = get_project_by_name(name)
+        if existing_project:
+            return pulumi.dynamic.CreateResult(
```

Comment:
Super super minor but I have nothing else to comment on so 🤷  - could compact this down to a single return statement by doing something like:

```python
# Check for existing project
project = get_project_by_name(name)
if not project:
    project = ProjectApi().create_project(...)
  
return pulumi.dynamic.CreateResult(...)
```

That way if the format of `id_` changes, it only needs to be updated in one spot.

Reply:
Yeah great idea, actioned!